suite/tests/api/ir_x86_4args_avx512_evex_mask_C.h

```diff
@@ -1959,3 +1959,11 @@ OPCODE(vsqrtsd_xhik7xhixhi, vsqrtsd, vsqrtsd_mask, X64_ONLY, REGARG(XMM16), REGA
        REGARG(XMM17), REGARG_PARTIAL(XMM18, OPSZ_8))
 OPCODE(vsqrtsd_xhik7xhimem, vsqrtsd, vsqrtsd_mask, X64_ONLY, REGARG(XMM16), REGARG(K7),
        REGARG(XMM17), MEMARG(OPSZ_8))
+OPCODE(vpdpbusd_zhik7zhizhi, vpdpbusd, vpdpbusd_mask, X64_ONLY, REGARG(ZMM0), REGARG(K0),
```

Comment:
This api.ir test gives us internal consistency between DR's decoder, encoder, and synthetic creation, but for ground-truth testing it would be great if we had test cases in our binutils comparison tests in third_party/binutils/test_decenc/.  Those were created several years ago and so  don't have these VNNI instructions, but it looks like the latest binutils does:
```
gas/testsuite/gas/i386/x86-64-avx512vnni_vl.s:  vpdpbusd        %xmm28, %xmm29, %xmm18   # AVX512{VNNI,VL}
```

I don't know that the transformation from the binutils test to the reformatted version in our third_party/ was automated but I think it's a relatively simple copy from the .s file?  So the request here is to add the various vnni test cases from the latest binutils into our third_party tests.

Reply:
This should probably be added as step 8) in that list.

Reply:
I will add these tests. Apart from the AVX512 tests in this file there are AVX-VNNI tests in avx-vnni.s. I'll add those too.

Reply:
I am seeing that the att syntax emitted by DR differs from binutils and the DR expect files seem to be aware of this difference.

Let's take this example in x86-64-evex-wig2.{d,s} in binutils
```
 15  +[a-f0-9]+:  62 f1 36 00 7b f0     vcvtusi2ss %eax,%xmm25,%xmm6
```
```
 10   vcvtusi2ss %eax, %xmm25, %xmm6
```

In drdecode_decenc_x86_64.expect `xmm25` and `eax` are swapped.
```
 62 f1 36 00 7b f0    vcvtusi2ss %xmm25, %eax, %xmm6
```
```
$ build/clients/bin64/drdisas -syntax att 62 f1 36 00 7b f0
 62 f1 36 00 7b f0    vcvtusi2ss %xmm25, %eax, %xmm6
```
Why is this so?

Reply:
One more example below. This shows the operand issue mentioned above and also another difference where DR output has `{%k0}`

```
avx512bw.d:[    ]*[a-f0-9]+:[   ]*62 f1 55 48 6b f4[    ]*vpackssdw %zmm4,%zmm5,%zmm6
drdecode_decenc_x86.expect: 62 f1 55 48 6b f4    vpackssdw %zmm5, %zmm4, %zmm6 {%k0}
```
It looks like the binutils representation is not used as-is and some processing is done to generate the expect files. Is that the case? Maybe there was some automation to make these changes? Or was some other method used to generate the assembly in the expect files?

Reply:
I see that DR does not print `{z}` or `{1to16}`

This is from binutils
```
avx512bw.d:[    ]*[a-f0-9]+:[   ]*62 f2 7d cf 1c f5[    ]*vpabsb %zmm5,%zmm6\{%k7\}\{z\}
avx512bw.d:[       ]*[a-f0-9]+:[   ]*62 f1 55 58 6b 30[    ]*vpackssdw \(%eax\)\{1to16\},%zmm5,%zmm6
```

This is from DR
```
drdecode_decenc_x86.expect: 62 f2 7d cf 1c f5    vpabsb %zmm5, %zmm6 {%k7}
drdecode_decenc_x86.expect: 62 f1 55 58 6b 30    vpackssdw %zmm5, (%eax), %zmm6 {%k0}
```

Reply:
The initial commit has some info: 4e99a45734d3ec97e388516596315e0e9dd7c829

Reply:
In binutils they have 8 files with VNNI tests (4 for x86 and 4 for x64) if we exclude the 4vnniw files. I took the x64 files and excluded `x86-64-avx-vnni-inval.s` which tests for unsupported instructions so I was left with the 3 `x86-64-avx*` files.
```
$ ls *vnni*s | grep -v 4vnniw
avx512vnni.s
avx512vnni_vl.s
avx-vnni-inval.s
avx-vnni.s
x86-64-avx512vnni.s
x86-64-avx512vnni_vl.s
x86-64-avx-vnni-inval.s
x86-64-avx-vnni.s
```
I took the opcodes from the `.d` files and converted them using DR disassembler. Then I manually compared the entries and  added the DR disassembly to the binutils expect file.

### x86-64-avx512vnni.s
The following set of instructions has 8 variants in this file - for 4 instructions x 2 syntax variants (att, intel). For brevity I have shown only 1 variant.

x86-64-avx512vnni.s
```
        vpdpwssd        %zmm17, %zmm18, %zmm18   # AVX512VNNI
        vpdpwssd        %zmm17, %zmm18, %zmm18{%k5}      # AVX512VNNI
        vpdpwssd        %zmm17, %zmm18, %zmm18{%k5}{z}   # AVX512VNNI
        vpdpwssd        0x123(%rax,%r14,8), %zmm18, %zmm18       # AVX512VNNI
        vpdpwssd        8128(%rdx), %zmm18, %zmm18       # AVX512VNNI Disp8
        vpdpwssd        508(%rdx){1to16}, %zmm18, %zmm18         # AVX512VNNI Disp8
```
x86-64-avx512vnni.d
```
[       ]*[a-f0-9]+:[   ]*62 a2 6d 40 52 d1[    ]*vpdpwssd %zmm17,%zmm18,%zmm18
[       ]*[a-f0-9]+:[   ]*62 a2 6d 45 52 d1[    ]*vpdpwssd %zmm17,%zmm18,%zmm18\{%k5\}
[       ]*[a-f0-9]+:[   ]*62 a2 6d c5 52 d1[    ]*vpdpwssd %zmm17,%zmm18,%zmm18\{%k5\}\{z\}
[       ]*[a-f0-9]+:[   ]*62 a2 6d 40 52 94 f0 23 01 00 00[     ]*vpdpwssd 0x123\(%rax,%r14,8\),%zmm18,%zmm18
[       ]*[a-f0-9]+:[   ]*62 e2 6d 40 52 52 7f[         ]*vpdpwssd 0x1fc0\(%rdx\),%zmm18,%zmm18
[       ]*[a-f0-9]+:[   ]*62 e2 6d 50 52 52 7f[         ]*vpdpwssd 0x1fc\(%rdx\)\{1to16\},%zmm18,%zmm18
```
Corresponding entries in drdecode_decenc_x86_64.expect
```
102556  62 a2 6d 40 52 d1    data16 vpdpwssd %zmm18, %zmm17, %zmm18 {%k0}
102557  62 a2 6d 45 52 d1    data16 vpdpwssd %zmm18, %zmm17, %zmm18 {%k5}
102558  62 a2 6d c5 52 d1    data16 vpdpwssd %zmm18, %zmm17, %zmm18 {%k5}
102559  62 a2 6d 40 52 94 f0 data16 vpdpwssd %zmm18, 0x00000123(%rax,%r14,8), %zmm18 {%k0}
102560  23 01 00 00
102561  62 e2 6d 40 52 52 7f data16 vpdpwssd %zmm18, 0x00001fc0(%rdx), %zmm18 {%k0}
102562  62 e2 6d 50 52 52 7f data16 vpdpwssd %zmm18, 0x000001fc(%rdx), %zmm18 {%k0}
```
### x86-64-avx512vnni_vl.s
Similarly for `x86-64-avx512vnni_vl.s` this set has 8 variants. The binutils and DR disassemblies are shown.
```
        vpdpwssd        %xmm20, %xmm22, %xmm26   # AVX512{VNNI,VL}
        vpdpwssd        %xmm20, %xmm22, %xmm26{%k3}      # AVX512{VNNI,VL}
        vpdpwssd        %xmm20, %xmm22, %xmm26{%k3}{z}   # AVX512{VNNI,VL}
        vpdpwssd        0x123(%rax,%r14,8), %xmm22, %xmm26       # AVX512{VNNI,VL}
        vpdpwssd        2032(%rdx), %xmm22, %xmm26       # AVX512{VNNI,VL} Disp8
        vpdpwssd        508(%rdx){1to4}, %xmm22, %xmm26  # AVX512{VNNI,VL} Disp8
        vpdpwssd        %ymm18, %ymm20, %ymm20   # AVX512{VNNI,VL}
        vpdpwssd        %ymm18, %ymm20, %ymm20{%k5}      # AVX512{VNNI,VL}
        vpdpwssd        %ymm18, %ymm20, %ymm20{%k5}{z}   # AVX512{VNNI,VL}
        vpdpwssd        0x123(%rax,%r14,8), %ymm20, %ymm20       # AVX512{VNNI,VL}
        vpdpwssd        4064(%rdx), %ymm20, %ymm20       # AVX512{VNNI,VL} Disp8
        vpdpwssd        508(%rdx){1to8}, %ymm20, %ymm20  # AVX512{VNNI,VL} Disp8
```
```
[       ]*[a-f0-9]+:[   ]*62 22 4d 00 52 d4[    ]*vpdpwssd %xmm20,%xmm22,%xmm26
[       ]*[a-f0-9]+:[   ]*62 22 4d 03 52 d4[    ]*vpdpwssd %xmm20,%xmm22,%xmm26\{%k3\}
[       ]*[a-f0-9]+:[   ]*62 22 4d 83 52 d4[    ]*vpdpwssd %xmm20,%xmm22,%xmm26\{%k3\}\{z\}
[       ]*[a-f0-9]+:[   ]*62 22 4d 00 52 94 f0 23 01 00 00[     ]*vpdpwssd 0x123\(%rax,%r14,8\),%xmm22,%xmm26
[       ]*[a-f0-9]+:[   ]*62 62 4d 00 52 52 7f[         ]*vpdpwssd 0x7f0\(%rdx\),%xmm22,%xmm26
[       ]*[a-f0-9]+:[   ]*62 62 4d 10 52 52 7f[         ]*vpdpwssd 0x1fc\(%rdx\)\{1to4\},%xmm22,%xmm26
[       ]*[a-f0-9]+:[   ]*62 a2 5d 20 52 e2[    ]*vpdpwssd %ymm18,%ymm20,%ymm20
[       ]*[a-f0-9]+:[   ]*62 a2 5d 25 52 e2[    ]*vpdpwssd %ymm18,%ymm20,%ymm20\{%k5\}
[       ]*[a-f0-9]+:[   ]*62 a2 5d a5 52 e2[    ]*vpdpwssd %ymm18,%ymm20,%ymm20\{%k5\}\{z\}
[       ]*[a-f0-9]+:[   ]*62 a2 5d 20 52 a4 f0 23 01 00 00[     ]*vpdpwssd 0x123\(%rax,%r14,8\),%ymm20,%ymm20
[       ]*[a-f0-9]+:[   ]*62 e2 5d 20 52 62 7f[         ]*vpdpwssd 0xfe0\(%rdx\),%ymm20,%ymm20
[       ]*[a-f0-9]+:[   ]*62 e2 5d 30 52 62 7f[         ]*vpdpwssd 0x1fc\(%rdx\)\{1to8\},%ymm20,%ymm20

```
```
102612  62 22 4d 00 52 d4    data16 vpdpwssd %xmm22, %xmm20, %xmm26 {%k0}
102613  62 22 4d 03 52 d4    data16 vpdpwssd %xmm22, %xmm20, %xmm26 {%k3}
102614  62 22 4d 83 52 d4    data16 vpdpwssd %xmm22, %xmm20, %xmm26 {%k3}
102615  62 22 4d 00 52 94 f0 data16 vpdpwssd %xmm22, 0x00000123(%rax,%r14,8), %xmm26 {%k0}
102616  23 01 00 00
102617  62 62 4d 00 52 52 7f data16 vpdpwssd %xmm22, 0x000007f0(%rdx), %xmm26 {%k0}
102618  62 62 4d 10 52 52 7f data16 vpdpwssd %xmm22, 0x000001fc(%rdx), %xmm26 {%k0}
102619  62 a2 5d 20 52 e2    data16 vpdpwssd %ymm20, %ymm18, %ymm20 {%k0}
102620  62 a2 5d 25 52 e2    data16 vpdpwssd %ymm20, %ymm18, %ymm20 {%k5}
102621  62 a2 5d a5 52 e2    data16 vpdpwssd %ymm20, %ymm18, %ymm20 {%k5}
102622  62 a2 5d 20 52 a4 f0 data16 vpdpwssd %ymm20, 0x00000123(%rax,%r14,8), %ymm20 {%k0}
102623  23 01 00 00
102624  62 e2 5d 20 52 62 7f data16 vpdpwssd %ymm20, 0x00000fe0(%rdx), %ymm20 {%k0}
102625  62 e2 5d 30 52 62 7f data16 vpdpwssd %ymm20, 0x000001fc(%rdx), %ymm20 {%k0}
```
### x86-64-avx-vnni.s
Similarly for `x86-64-avx-vnni.s`
```
.macro test_insn mnemonic
        \mnemonic        %xmm12, %xmm4, %xmm2
        {evex} \mnemonic %xmm12, %xmm4, %xmm2
        {vex}  \mnemonic %xmm12, %xmm4, %xmm2
        {vex3} \mnemonic %xmm12, %xmm4, %xmm2
        {vex}  \mnemonic (%rcx), %xmm4, %xmm2
        {vex3} \mnemonic (%rcx), %xmm4, %xmm2
        \mnemonic        %xmm22, %xmm4, %xmm2
.endm

        .text
_start:
        test_insn vpdpbusd
        test_insn vpdpwssd
        test_insn vpdpbusds
        test_insn vpdpwssds

        .arch .avx_vnni
        vpdpbusd        %xmm12, %xmm4, %xmm2
```
First set of entries in `x86-64-avx-vnni.d`
```
 +[a-f0-9]+:    62 d2 5d 08 50 d4       vpdpbusd %xmm12,%xmm4,%xmm2
 +[a-f0-9]+:    62 d2 5d 08 50 d4       vpdpbusd %xmm12,%xmm4,%xmm2
 +[a-f0-9]+:    c4 c2 59 50 d4          \{vex\} vpdpbusd %xmm12,%xmm4,%xmm2
 +[a-f0-9]+:    c4 c2 59 50 d4          \{vex\} vpdpbusd %xmm12,%xmm4,%xmm2
 +[a-f0-9]+:    c4 e2 59 50 11          \{vex\} vpdpbusd \(%rcx\),%xmm4,%xmm2
 +[a-f0-9]+:    c4 e2 59 50 11          \{vex\} vpdpbusd \(%rcx\),%xmm4,%xmm2
 +[a-f0-9]+:    62 b2 5d 08 50 d6       vpdpbusd %xmm22,%xmm4,%xmm2
```
Corresponding DR disassembly
```
102724  62 d2 5d 08 50 d4    data16 vpdpbusd %xmm4, %xmm12, %xmm2 {%k0}
102725  62 d2 5d 08 50 d4    data16 vpdpbusd %xmm4, %xmm12, %xmm2 {%k0}
102726  c4 c2 59 50 d4       data16 vpdpbusd %xmm2, %xmm4, %xmm12
102727  c4 c2 59 50 d4       data16 vpdpbusd %xmm2, %xmm4, %xmm12
102728  c4 e2 59 50 11       data16 vpdpbusd %xmm2, %xmm4, (%rcx)
102729  c4 e2 59 50 11       data16 vpdpbusd %xmm2, %xmm4, (%rcx)
102730  62 b2 5d 08 50 d6    data16 vpdpbusd %xmm4, %xmm22, %xmm2 {%k0}
```

I ignored these differences

- operand ordering in DR att disassembly
- extra `{%k0}` in DR disassembly
- missing `{z}` and `{1to16}` etc in DR disassembly
- missing `{vex}` prefix in DR disassembly

Reply:
I added a "step 8" in the comments. Do we need anything else to resolve this conversation?

Reply:
No seems fine.  I guess I assumed the author would resolve; that is the convention in other code review systems: I guess our docs don't make it clear.